upstream.go

```diff
@@ -155,6 +162,84 @@ func (uc *upstreamConn) isChannel(entity string) bool {
 	return false
 }
 
+func (uc *upstreamConn) getPendingList() *pendingList {
+	uc.user.pendingListsLock.Lock()
+	defer uc.user.pendingListsLock.Unlock()
+	for _, pl := range uc.user.pendingLists {
+		if _, ok := pl.pendingRequests[uc.network.ID]; !ok {
+			continue
+		}
+		return &pl
+	}
+	return nil
+}
+
+func (uc *upstreamConn) endPendingLists(all bool) (found bool) {
+	found = false
+	uc.user.pendingListsLock.Lock()
+	defer uc.user.pendingListsLock.Unlock()
+	for i := 0; i < len(uc.user.pendingLists); i++ {
+		pl := uc.user.pendingLists[i]
+		if _, ok := pl.pendingRequests[uc.network.ID]; !ok {
+			continue
+		}
+		delete(pl.pendingRequests, uc.network.ID)
+		if len(pl.pendingRequests) == 0 {
+			uc.user.pendingLists = append(uc.user.pendingLists[:i], uc.user.pendingLists[i+1:]...)
+			i--
+			uc.forEachDownstreamByID(pl.downstreamID, func(dc *downstreamConn) {
+				dc.SendMessage(&irc.Message{
+					Prefix:  dc.srv.prefix(),
+					Command: irc.RPL_LISTEND,
+					Params:  []string{dc.nick, "End of /LIST"},
+				})
+			})
+		}
+		found = true
+		if !all {
+			delete(uc.pendingListDownstreamSet, pl.downstreamID)
+			uc.user.forEachUpstream(func(uc *upstreamConn) {
```

Comment:
Since this function can only "unblock" a LIST command for the current upstream connection `uc`, do we really need to iterate over all upstream connections here?

Reply:
Suppose we're connected to downstreams *a*, *b*; and the client does LIST twice.

Pending now schematically looks like this: `[[a, b], [a, b]]`. Now imagine *a* answers much faster than *b*. After *a* answers, we get: `[[b], [a, b]]`. We can't yet send the other LIST to *a* because it would be sent before the first RPL_LISTEND. When *b* finally replies, we get: `[[], [a, b]]`, ie `[[a, b]]`. We now need to resend both  `a` and `b` for the next LIST.

Hence the need to iterate over all upstreams.